File: src/game/rng.ts
/** Deterministic 32-bit PRNG. Returns floats in [0, 1). */
export function mulberry32(seed: number): () => number {
  let a = seed >>> 0
  return () => {
    a = (a + 0x6d2b79f5) >>> 0
    let t = a
    t = Math.imul(t ^ (t >>> 15), t | 1)
    t ^= t + Math.imul(t ^ (t >>> 7), t | 61)
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296
  }
}

/** Mix two integers into one seed. */
export function hashSeed(a: number, b: number): number {
  return (Math.imul(a ^ 0x9e3779b9, 0x85ebca6b) ^ Math.imul(b + 0x7f4a7c15, 0xc2b2ae35)) >>> 0
}

File: server/evomaze/rng.py
"""Deterministic 32-bit PRNG (mulberry32) so every maze is reproducible from its seed."""

_M32 = 0xFFFFFFFF


def _imul(a: int, b: int) -> int:
    return (a * b) & _M32


def mulberry32(seed: int):
    a = seed & _M32

    def rng() -> float:
        nonlocal a
        a = (a + 0x6D2B79F5) & _M32
        t = a
        t = _imul(t ^ (t >> 15), t | 1)
        t ^= (t + _imul(t ^ (t >> 7), t | 61)) & _M32
        return ((t ^ (t >> 14)) & _M32) / 4294967296

    return rng


def hash_seed(a: int, b: int) -> int:
    """Mix two integers into one seed."""
    return (_imul((a ^ 0x9E3779B9) & _M32, 0x85EBCA6B) ^ _imul((b + 0x7F4A7C15) & _M32, 0xC2B2AE35)) & _M32
